### DISCUSSION TITLE
App path is hardcoded, making it (seemingly) impossible to have Livewire files in a package

### DISCUSSION THREAD

**Discussion by gavinhewitt** [livewire/livewire#3415] [1. Help]
I've created a private package that uses Livewire and I'm trying to write tests for it (testing with Orchestra Testbench). I have most of it set up correctly (I think) but I keep on getting 
```
Symfony\Component\Finder\Exception\DirectoryNotFoundException: The "/pkg-name/vendor/orchestra/testbench-core/laravel/app/Http/Livewire/" directory does not exist. 
```

Technically this correct because the Livewire files are located in `VENDOR_NAME/PACKAGE_NAME/src/Http/Livewire`. However, I can't seem to change the location, probably because of this function in ComponentParser.php which effectively sets the app path to `/pkg-name/vendor/orchestra/testbench-core/laravel/app/Http/Livewire/`:

```
public static function generatePathFromNamespace($namespace)
{
        $name = str($namespace)->finish('\\')->replaceFirst(app()->getNamespace(), '');
        return app('path').'/'.str_replace('\\', '/', $name);
}
```

Is there any other way to point to the right location?

**Comment by joshhanley** [comment]
@gavinhewitt you need to manually register your components in your package service provider for them to work properly. 

I have a demo Livewire package setup, which you can have a look at how I am registering the components, here is the link to the package's service provider https://github.com/joshhanley/livewire-package-demo/blob/main/src/LivewirePackageDemoServiceProvider.php

Hope this helps!

**Reply by gavinhewitt** [reply]
Thanks for your response @joshhanley !

However, this is exactly what I have done. I'm gonna look at your code a bot more thorough but at first glance it looks like my setup :-(

**Reply by joshhanley** [reply]
@gavinhewitt yeah the other thing to check out is the `tests/Browser/TestCase.php` file. It might be a little different as I'm using dusk to test with (using Livewire Dusk Testbench) but you will see the service provider is registered for the testbench to use.

I can't remember if you have to register package service providers in the tests when use the standard TestBench, but I suspect you do, to ensure the components are registered and available in the tests.

Also check your autoloading in `composer.json`.

**Comment by ju5t** [comment]
@joshhanley sorry to tag you again, but I'm having the same issue and I have no clue where to look for.

Our package provides components that are meant to be extended. So our `PackageComponent` lives in `src` and the version that we test is in `test`. I'm not keen on adding components specifically made for testing to our ServiceProvider, but when trying it, it didn't make a difference either.

And although our base component isn't meant to be used on its own, adding that to our ServiceProvider got me the same result. Both returned an error as the one from @gavinhewitt. 

I have went through some of the packages I know to include Livewire components but I couldn't find anything that they do different compared to our package. 

Is there anything that has maybe changed since last year that made it easier to test Livewire from within a package?

**Reply by joshhanley** [reply]
@ju5t sorry for the late reply!

Yeah there is a PR #4923 that has been merged that will be released in the next version that fixes that issue.

Hope this helps!
